crates/apollo-compiler/src/diagnostic.rs

```diff
@@ -0,0 +1,304 @@
+                }
+            }
+            let source_file = self.0.get(file_id)?;
+            Some(Box::new(Path(source_file.clone())))
+        } else {
+            struct NoSourceFile;
+            impl fmt::Display for NoSourceFile {
+                fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                    f.write_str("(no source file)")
+                }
+            }
+            Some(Box::new(NoSourceFile))
+        }
+    }
+}
+
+/// A pretty-printable diagnostic.
+#[derive(Debug)]
+pub struct Diagnostic<T> {
+    pub sources: SourceMap,
+    pub location: Option<NodeLocation>,
+    pub error: T,
+}
+
+impl<T> std::ops::Deref for Diagnostic<T> {
+    type Target = T;
+    fn deref(&self) -> &Self::Target {
+        &self.error
+    }
+}
+
+impl<T: std::error::Error + ToDiagnostic> std::error::Error for Diagnostic<T> {}
+
+impl<T: ToDiagnostic> ToDiagnostic for &T {
+    fn location(&self) -> Option<NodeLocation> {
+        ToDiagnostic::location(*self)
+    }
+
+    fn report(&self, report: &mut DiagnosticReport) {
+        ToDiagnostic::report(*self, report)
+    }
+}
+
+impl<T: ToDiagnostic> Diagnostic<T> {
+    /// Write the report to a [`Write`], with colors.
+    ///
+    /// If colored output is not desired, consider wrapping the [`Write`] with [`anstream`].
+    ///
+    /// [`Write`]: std::io::Write
+    pub fn write(&self, w: impl std::io::Write) -> std::io::Result<()> {
+        let mut report = DiagnosticReport::builder(self.sources.clone(), self.error.location());
+        self.error.report(&mut report);
+        report.write(w)
+    }
+}
+
+impl<T: ToDiagnostic> fmt::Display for Diagnostic<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        struct StripColorAdaptor<'a, 'b> {
+            f: &'a mut fmt::Formatter<'b>,
+            strip: anstream::adapter::StripBytes,
+        }
+        impl io::Write for StripColorAdaptor<'_, '_> {
+            fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+                for printable in self.strip.strip_next(buf) {
+                    let s = std::str::from_utf8(printable).map_err(|_| io::ErrorKind::Other)?;
+                    self.f.write_str(s).map_err(|_| io::ErrorKind::Other)?;
+                }
+
+                Ok(buf.len())
+            }
+
+            fn flush(&mut self) -> io::Result<()> {
+                Ok(())
+            }
+        }
+
+        struct ColorAdaptor<'a, 'b> {
+            f: &'a mut fmt::Formatter<'b>,
+        }
+        impl io::Write for ColorAdaptor<'_, '_> {
+            fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+                let s = std::str::from_utf8(buf).map_err(|_| io::ErrorKind::Other)?;
+                self.f.write_str(s).map_err(|_| io::ErrorKind::Other)?;
+                Ok(buf.len())
+            }
+
+            fn flush(&mut self) -> io::Result<()> {
+                Ok(())
+            }
+        }
+
+        if f.alternate() {
```

Comment:
I came up with this use of `alternate` but i don’t like it that much. What if instead we had `Debug` enable colors (for printed panic messages) and `Display` disable them (for `.to_string()` instead of `.to_string_no_color()`)?